### ISSUE TITLE
Race condition between printing server output and killing server process

### ISSUE THREAD

**Issue by yume-chan** [Genymobile/scrcpy#1992]
- [x] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [x] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: Windows 10 x64
 - scrcpy version: Latest dev branch
 - installation method: manual build
 - device model: Samsung Galaxy S9
 - Android version: 10

**Describe the bug**

I'm working on a third-party client implementation, and I want to parse available encoders (#1827) programmatically.

I have known that the server process won't exit by its own, I have to kill it somehow.

However, I can't find out when should I do it.

---

Looking at your implementation, it looks like it's pure luck that the server printed all error message before the client killed it.

As an example, normally the client will output like this:

```
PS C\:<redacted>> .\scrcpy.exe -V debug --encoder _
INFO: scrcpy 1.16 <https://github.com/Genymobile/scrcpy>
DEBUG: Using server (portable): C:\<redacted>\scrcpy-server
C:\<redacted>\scrcpy-server:... file pushed, 0 skipped. 69.8 MB/s (34482 bytes in 0.000s)
DEBUG: Screensaver enabled
[server] INFO: Device: samsung SM-G9600 (Android 10)
[server] DEBUG: Creating encoder by name: '_'
DEBUG: Starting stream thread
DEBUG: Starting controller thread
DEBUG: Starting receiver thread
DEBUG: End of frames
DEBUG: Receiver stopped
[server] DEBUG: Controller stopped
[server] ERROR: Exception on thread Thread[main,5,main]
com.genymobile.scrcpy.InvalidEncoderException: There is no encoder having name '_"
        at com.genymobile.scrcpy.ScreenEncoder.createCodec(ScreenEncoder.java:176)
        at com.genymobile.scrcpy.ScreenEncoder.internalStreamScreen(ScreenEncoder.java:77)
        at com.genymobile.scrcpy.ScreenEncoder.streamScreen(ScreenEncoder.java:60)
        at com.genymobile.scrcpy.Server.scrcpy(Server.java:78)
        at com.genymobile.scrcpy.Server.main(Server.java:239)
        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:342)
[server] ERROR: Try to use one of the available encoders:
[server] ERROR:     scrcpy --encoder-name 'OMX.qcom.video.encoder.avc'
[server] ERROR:     scrcpy --encoder-name 'c2.android.avc.encoder'
[server] ERROR:     scrcpy --encoder-name 'OMX.google.h264.encoder'
INFO: Renderer: direct3d
DEBUG: Trilinear filtering disabled (not an OpenGL renderer)
INFO: Initial texture: 1440x2960
DEBUG: Video stream stopped
WARN: Device disconnected
DEBUG: quit...
DEBUG: Server terminated
Press Enter to continue...
```

I found that just after `DEBUG: End of frames` got printed, you sent a SDL event to kill the server.

https://github.com/Genymobile/scrcpy/blob/d0f5a7fd9f49e3780c40f34eeeaa1a6574558a47/app/src/stream.c#L243

https://github.com/Genymobile/scrcpy/blob/d0f5a7fd9f49e3780c40f34eeeaa1a6574558a47/app/src/stream.c#L267

https://github.com/Genymobile/scrcpy/blob/d0f5a7fd9f49e3780c40f34eeeaa1a6574558a47/app/src/stream.c#L65-L70

However, the event didn't get processed until the renderer initialization routine was done.

https://github.com/Genymobile/scrcpy/blob/76c2c6e69dc074031a81c4e050347e374cbfb776/app/src/scrcpy.c#L421-L428

https://github.com/Genymobile/scrcpy/blob/76c2c6e69dc074031a81c4e050347e374cbfb776/app/src/scrcpy.c#L447-L448

https://github.com/Genymobile/scrcpy/blob/76c2c6e69dc074031a81c4e050347e374cbfb776/app/src/scrcpy.c#L468-L469

if I skip the renderer initialization with `--no-display` (`-N`), I only got this

```
PS C:\<redacted>> .\scrcpy.exe -V debug -N -r 1.mp4 --encoder _
INFO: scrcpy 1.16 <https://github.com/Genymobile/scrcpy>
DEBUG: Using server (portable): C:\<redacted>\scrcpy-server
C:\<redacted>\scrcpy-server:... file pushed, 0 skipped. 43.4 MB/s (34482 bytes in 0.001s)
[server] INFO: Device: samsung SM-G9600 (Android 10)
[server] DEBUG: Creating encoder by name: '_'
DEBUG: Starting stream thread
INFO: Recording started to mp4 file: 1.mp4
DEBUG: Starting recorder thread
DEBUG: End of frames
INFO: Finishing recording...
DEBUG: Recorder thread ended
ERROR: Recording failed to 1.mp4
[server] DEBUG: Controller stopped
DEBUG: Video stream stopped
WARN: Device disconnected
DEBUG: quit...
DEBUG: Server terminated
Press Enter to continue...
```

Now the error message is gone.

**Comment by rom1v** [comment]
> Looking at your implementation, it looks like it's pure luck that the server printed all error message before the client killed it.

Yes, this is right.

It was done that way to avoid keeping a running (blocked) process on the device side in any case (due to MediaCodec deadlock on some devices, or other blocking calls for whatever reason).

The server error reporting (especially for listing encoders) is very hacky (it just logs the data), this is useful for debugging (it's still available in `adb logcat`) but not very friendly.

**Comment by yume-chan** [comment]
So it's not expected to be changed, for example by printing error message before closing the video socket?

Now my implementation starts the server (with encoder name `_`), wait 1 second, then kill the server. I think I can live with that.

**Comment by rom1v** [comment]
> So it's not expected to be changed

Yes, it might, but I "planned" to do some refactors, in particular to start the server from a separate thread (and also in the long term separate the "core" and the SDL UI). I keep your issue in mind. Thank you.

**Comment by rom1v** [comment]
Could you please test [`racecond_on_close`](https://github.com/Genymobile/scrcpy/commits/racecond_on_close) please, it should fix the issue.

**Comment by yume-chan** [comment]
Happy new year!

I'm on vacation now and my dev machine is not with me. Will definitely try that once back at work.

**Comment by rom1v** [comment]
:+1:

I wrote a second version: [`racecond_on_close.2`](https://github.com/Genymobile/scrcpy/commits/racecond_on_close), which lets the server terminate properly for 2 seconds, then kill it.

The problem was that on some devices, closing the sockets is not sufficient while the device is asleep, so the server was not terminated (that's why I always killed it). But now it will let it terminate properly, unless it is blocked for more than 2 seconds.

**Comment by rom1v** [comment]
Merged in `master` and released (v1.17):
 - 90f835663004d9dc4969a76a1ef2641af9b02633
 - 10b749e27d34f91d04ff165134fda23e6ec2cc3b
